File: holoken-core/src/main/java/com/holokenmod/game/SaveGame.java
package com.holokenmod.game;

import com.holokenmod.grid.Grid;
import com.holokenmod.grid.GridCage;
import com.holokenmod.grid.GridCageAction;
import com.holokenmod.grid.GridCell;
import com.holokenmod.grid.GridSize;

import org.apache.commons.io.FileUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileWriter;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Optional;

public class SaveGame {
	public static final String SAVEGAME_AUTO_NAME = "autosave";
	public static final String SAVEGAME_NAME_PREFIX_ = "savegame_";
	
	private final Logger LOGGER = LoggerFactory.getLogger(SaveGame.class);
	
	private final File filename;

	private SaveGame(File filename) {
		this.filename = filename;
	}
	
	public static SaveGame createWithDirectory(final File directory) {
		return new SaveGame(getAutosave(directory));
	}
	
	public static SaveGame createWithFile(final File filename) {
		return new SaveGame(filename);
	}
	
	public void Save(final Grid grid) {
		try (final BufferedWriter writer = new BufferedWriter(new FileWriter(this.filename))) {
			final long now = System.currentTimeMillis();
			writer.write(now + "\n");
			writer.write(grid.getGridSize() + "\n");
			writer.write(grid.getPlayTime() + "\n");
			writer.write(grid.isActive() + "\n");
			for (final GridCell cell : grid.getCells()) {
				writer.write("CELL:");
				writer.write(cell.getCellNumber() + ":");
				writer.write(cell.getRow() + ":");
				writer.write(cell.getColumn() + ":");
				writer.write(cell.getCageText() + ":");
				writer.write(cell.getValue() + ":");
				writer.write(cell.getUserValue() + ":");
				for (final int possible : cell.getPossibles()) {
					writer.write(possible + ",");
				}
				writer.write("\n");
			}
			if (grid.getSelectedCell() != null) {
				writer.write("SELECTED:" + grid.getSelectedCell()
						.getCellNumber() + "\n");
			}
			final ArrayList<GridCell> invalidchoices = grid.invalidsHighlighted();
			if (invalidchoices.size() > 0) {
				writer.write("INVALID:");
				for (final GridCell cell : invalidchoices) {
					writer.write(cell.getCellNumber() + ",");
				}
				writer.write("\n");
			}
			final ArrayList<GridCell> cheatedcells = grid.cheatedHighlighted();
			if (cheatedcells.size() > 0) {
				writer.write("CHEATED:");
				for (final GridCell cell : cheatedcells) {
					writer.write(cell.getCellNumber() + ",");
				}
				writer.write("\n");
			}
			for (final GridCage cage : grid.getCages()) {
				writer.write("CAGE:");
				writer.write(cage.getId() + ":");
				writer.write(cage.getAction().name() + ":");
				writer.write("NOTHING" + ":");
				writer.write(cage.getResult() + ":");
				writer.write(cage.getCellNumbers());
				//writer.write(":" + cage.isOperatorHidden());
				writer.write("\n");
			}
		} catch (final IOException e) {
			LOGGER.debug("Error saving game: " + e.getMessage());
			return;
		}
		LOGGER.debug("Saved game.");
	}
	
	public long ReadDate() {
		try (final InputStream ins = new FileInputStream((this.filename));
			 final BufferedReader br = new BufferedReader(new InputStreamReader(ins), 8192)) {
			return Long.parseLong(br.readLine());
		} catch (final NumberFormatException | IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
		return 0;
	}
	
	public Optional<Grid> restore() {
		String line;
		BufferedReader br = null;
		InputStream ins = null;
		String[] cellParts;
		String[] cageParts;
		
		if (this.filename.length() == 0) {
			return Optional.empty();
		}
		
		try {
			LOGGER.info("test " + this.filename.getAbsolutePath() + " - " + this.filename.length());
			LOGGER.info("savefile " + FileUtils.readFileToString(this.filename));
			
			ins = new FileInputStream((this.filename));
			br = new BufferedReader(new InputStreamReader(ins), 8192);
			long creationDate = Long.parseLong(br.readLine());
			
			final String gridSizeString = br.readLine();
			GridSize gridSize = GridSize.create(gridSizeString);
			
			long playTime = Long.parseLong(br.readLine());
			
			final Grid grid = new Grid(gridSize, creationDate);
			
			grid.setActive(br.readLine().equals("true"));
			grid.setPlayTime(playTime);
			
			while ((line = br.readLine()) != null) {
                if (!line.startsWith("CELL:")) {
                    break;
                }
				cellParts = line.split(":");
				
				final int cellNum = Integer.parseInt(cellParts[1]);
				final int row = Integer.parseInt(cellParts[2]);
				final int column = Integer.parseInt(cellParts[3]);
				
				final GridCell cell = new GridCell(cellNum, row, column);
				
				cell.setCagetext(cellParts[4]);
				cell.setValue(Integer.parseInt(cellParts[5]));
				cell.setUserValue(Integer.parseInt(cellParts[6]));
                if (cellParts.length == 8) {
                    for (final String possible : cellParts[7].split(",")) {
                        cell.addPossible(Integer.parseInt(possible));
                    }
                }
				grid.addCell(cell);
			}
			if (line.startsWith("SELECTED:")) {
				final int selected = Integer.parseInt(line.split(":")[1]);
				grid.setSelectedCell(grid.getCell(selected));
				grid.getCell(selected).setSelected(true);
				line = br.readLine();
			}
			if (line.startsWith("INVALID:")) {
				final String invalidlist = line.split(":")[1];
				for (final String cellId : invalidlist.split(",")) {
					final int cellNum = Integer.parseInt(cellId);
					final GridCell c = grid.getCell(cellNum);
					c.setSelected(true);
				}
				line = br.readLine();
			}
			if (line.startsWith("CHEATED")) {
				final String cheatedlist = line.split(":")[1];
				for (final String cellId : cheatedlist.split(",")) {
					final int cellNum = Integer.parseInt(cellId);
					final GridCell c = grid.getCell(cellNum);
					c.setCheated(true);
				}
				line = br.readLine();
			}
			do {
				cageParts = line.split(":");
				final GridCage cage;
				cage = new GridCage(grid);
				cage.setCageId(Integer.parseInt(cageParts[1]));
				cage.setAction(GridCageAction.valueOf(cageParts[2]));
				cage.setResult(Integer.parseInt(cageParts[4]));
				for (final String cellId : cageParts[5].split(",")) {
					final int cellNum = Integer.parseInt(cellId);
					final GridCell c = grid.getCell(cellNum);
					c.setCage(cage);
					cage.addCell(c);
				}
				grid.getCages().add(cage);
			} while ((line = br.readLine()) != null);
			
			return Optional.of(grid);
		} catch (final IOException e) {
			LOGGER.info(e.getMessage(), e);
			
			return Optional.empty();
		} catch (final Exception e) {
			LOGGER.error(e.getMessage(), e);
			
			return Optional.empty();
		} finally {
			try {
				if (ins != null) {
					ins.close();
				}
				
				if (br != null) {
					br.close();
				}
                if (this.filename.getCanonicalPath().equals(getAutosave(this.filename.getParentFile()))) {
                    filename.delete();
                }
			} catch (final Exception ignored) {
			}
		}
	}
	
	private static File getAutosave(File directory) {
		return new File(directory, SAVEGAME_AUTO_NAME);
	}
}

File: holoken-core/src/main/kotlin/com/holokenmod/game/SaveGame.kt
package com.holokenmod.game

import com.holokenmod.creation.cage.GridCageType
import com.holokenmod.grid.Grid
import com.holokenmod.grid.GridCage
import com.holokenmod.grid.GridCageAction
import com.holokenmod.grid.GridCell
import com.holokenmod.grid.GridSize
import com.holokenmod.options.CurrentGameOptionsVariant
import com.holokenmod.options.GameVariant
import mu.KotlinLogging
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.File
import java.io.FileInputStream
import java.io.FileWriter
import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader

private val logger = KotlinLogging.logger {}

class SaveGame private constructor(private val filename: File) {
    fun save(grid: Grid) {
        try {
            BufferedWriter(FileWriter(filename)).use { writer ->
                val now = System.currentTimeMillis()
                writer.write(
                    """
    $now
    
                    """.trimIndent()
                )
                writer.write(
                    """
    ${grid.gridSize}
    
                    """.trimIndent()
                )
                writer.write(
                    """
    ${grid.playTime}
    
                    """.trimIndent()
                )
                writer.write(
                    """
    ${grid.isActive}
    
                    """.trimIndent()
                )
                for (cell in grid.cells) {
                    writer.write("CELL:")
                    writer.write(cell.cellNumber.toString() + ":")
                    writer.write(cell.row.toString() + ":")
                    writer.write(cell.column.toString() + ":")
                    writer.write(cell.value.toString() + ":")
                    writer.write(cell.userValue.toString() + ":")
                    for (possible in cell.possibles) {
                        writer.write("$possible,")
                    }
                    writer.write("\n")
                }
                grid.selectedCell?.let {
                    writer.write(
                        "SELECTED:" + it.cellNumber + "\n"
                    )
                }
                val invalidchoices = grid.invalidsHighlighted()
                if (invalidchoices.isNotEmpty()) {
                    writer.write("INVALID:")
                    invalidchoices.forEach {
                        writer.write(it.cellNumber.toString() + ",")
                    }
                    writer.write("\n")
                }
                val cheatedcells = grid.cheatedHighlighted()
                if (cheatedcells.isNotEmpty()) {
                    writer.write("CHEATED:")
                    cheatedcells.forEach {
                        writer.write(it.cellNumber.toString() + ",")
                    }
                    writer.write("\n")
                }
                grid.cages.forEach {
                    writer.write("CAGE:")
                    writer.write(it.id.toString() + ":")
                    writer.write(it.action.name + ":")
                    writer.write(it.cageType.name + ":")
                    writer.write(it.result.toString() + ":")
                    writer.write(it.cellNumbers)
                    // writer.write(":" + cage.isOperatorHidden());
                    writer.write("\n")
                }
            }
        } catch (e: IOException) {
            logger.error { "Error saving game: " + e.message }
            return
        }
        logger.debug { "Saved game." }
    }

    fun readDate(): Long {
        try {
            FileInputStream(filename).use { ins ->
                BufferedReader(
                    InputStreamReader(ins),
                    8192
                ).use { br -> return br.readLine().toLong() }
            }
        } catch (e: NumberFormatException) {
            // TODO Auto-generated catch block
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        }
        return 0
    }

    fun restore(): Grid? {
        var br: BufferedReader? = null
        var ins: InputStream? = null
        var cageParts: Array<String>
        return if (filename.length() == 0L) {
            null
        } else {
            try {
                logger.info { "test " + filename.absolutePath + " - " + filename.length() }
                logger.info { "savefile " + filename.readText() }
                ins = FileInputStream(filename)
                br = BufferedReader(InputStreamReader(ins), 8192)
                val creationDate = br.readLine().toLong()
                val gridSizeString = br.readLine()
                val gridSize: GridSize = GridSize.create(gridSizeString)
                val playTime = br.readLine().toLong()

                // TODO: Load and Save correct GameOptionsVariant
                val variant = GameVariant(
                    gridSize,
                    CurrentGameOptionsVariant.instance
                )
                val grid = Grid(variant, creationDate)
                grid.isActive = br.readLine() == "true"
                grid.playTime = playTime

                var line = readCells(br, grid)

                if (line.startsWith("SELECTED:")) {
                    val selected = line.split(":").dropLastWhile { it.isEmpty() }
                        .toTypedArray()[1].toInt()
                    grid.selectedCell = grid.getCell(selected)
                    grid.getCell(selected).isSelected = true
                    line = br.readLine()
                }
                if (line.startsWith("INVALID:")) {
                    val invalidlist = line.split(":").dropLastWhile { it.isEmpty() }
                        .toTypedArray()[1]
                    for (cellId in invalidlist.split(",").dropLastWhile { it.isEmpty() }
                        .toTypedArray()) {
                        val cellNum = cellId.toInt()
                        val c = grid.getCell(cellNum)
                        c.isSelected = true
                    }
                    line = br.readLine()
                }
                if (line.startsWith("CHEATED")) {
                    val cheatedlist = line.split(":").dropLastWhile { it.isEmpty() }
                        .toTypedArray()[1]
                    for (cellId in cheatedlist.split(",").dropLastWhile { it.isEmpty() }
                        .toTypedArray()) {
                        val cellNum = cellId.toInt()
                        val c = grid.getCell(cellNum)
                        c.isCheated = true
                    }
                    line = br.readLine()
                }

                var rawLine: String? = line

                do {
                    val currentLine = rawLine as String

                    cageParts = currentLine.split(":").dropLastWhile { it.isEmpty() }.toTypedArray()
                    val cage = GridCage(cageParts[1].toInt(), grid, GridCageAction.valueOf(cageParts[2]), GridCageType.valueOf(cageParts[3]))
                    cage.result = cageParts[4].toInt()
                    for (cellId in cageParts[5].split(",").dropLastWhile { it.isEmpty() }
                        .toTypedArray()) {
                        val cellNum = cellId.toInt()
                        val c = grid.getCell(cellNum)
                        c.cage = cage
                        cage.addCell(c)
                    }
                    grid.cages = grid.cages + cage
                } while (br.readLine().also { rawLine = it } != null)

                grid.setCageTexts()

                return grid
            } catch (e: IOException) {
                logger.info(e.message, e)
                return null
            } catch (e: Exception) {
                logger.error(e.message, e)
                return null
            } finally {
                try {
                    ins?.close()
                    br?.close()
                } catch (ignored: Exception) {
                }
            }
        }
    }

    private fun readCells(
        br: BufferedReader,
        grid: Grid
    ): String {
        var rawLine: String?

        while (br.readLine().also { rawLine = it } != null) {
            val line = rawLine as String

            if (!line.startsWith("CELL:")) {
                return line
            }
            val cellParts = line.split(":").dropLastWhile { it.isEmpty() }.toTypedArray()
            val cellNum = cellParts[1].toInt()
            val row = cellParts[2].toInt()
            val column = cellParts[3].toInt()
            val cell = GridCell(grid, cellNum, row, column)
            cell.value = cellParts[4].toInt()
            cell.userValue = cellParts[5].toInt()
            if (cellParts.size == 7) {
                for (possible in cellParts[6].split(",")
                    .dropLastWhile { it.isEmpty() }.toTypedArray()) {
                    cell.addPossible(possible.toInt())
                }
            }
            grid.addCell(cell)
        }

        return ""
    }

    companion object {
        const val SAVEGAME_AUTO_NAME = "autosave"
        const val SAVEGAME_NAME_PREFIX_ = "savegame_"

        @JvmStatic
        fun createWithDirectory(directory: File): SaveGame {
            return SaveGame(getAutosave(directory))
        }

        @JvmStatic
        fun createWithFile(filename: File): SaveGame {
            return SaveGame(filename)
        }

        private fun getAutosave(directory: File): File {
            return File(directory, SAVEGAME_AUTO_NAME)
        }
    }
}
